functions/src/pdfBuilder/donationCertificateBuilderFunction.ts

```diff
@@ -0,0 +1,140 @@
+import * as functions from 'firebase-functions';
+//import { createAnduploadDonationCertificate } from './buildDonationCertificateCH';
+import { initializeApp } from "firebase/app";
+import { getFirestore, connectFirestoreEmulator } from 'firebase/firestore';
+import { doc, setDoc, getDocs, collection } from "firebase/firestore";
+import { createAndUploadDonationCertificate } from './generateDonationCertificatePDF';
+import * as config from '../config';
+
+let FIRESTORE_EMULATOR_REQUIRED = true;
+
+export const bulkDonationCertificateBuilderFunction = functions.https.onCall(async (data, context) => {
+	const app = await initializeFirebaseApp();
+	const firestore = await initializeFirestore(app);
+	const now = new Date();
+	const currentDate = now.toLocaleDateString('de-DE');
+
+	const calculationYear = data.year
+
+	for (let i = 0; i < data.users.length; ++i) {
+		await createAndUploadDonationCertificate(await prepareData(data.users[i], calculationYear, currentDate, firestore), app, firestore);
+		//await addDocumentToUser(data.users[i], calculationYear, downloadUrl, firestore);
+	};
+
+	const responseString = `Donation Certificates for ${data.users.length} users created for the year ${calculationYear}`;
+	
+	return responseString;
+	
+});
+
+export const initializeFirebaseApp = async () => {
+	var firebaseConfig = {
+		apiKey: '0',
+		authDomain: 'demo-social-income.firebaseapp.com',
+		projectId: 'demo-social-income-prod',
+		storageBucket: 'demo-social-income.appspot.com',
+		messagingSenderId: '0',
+		appId: '0',
+		measurementId: '0',
+	};
+	
+	return initializeApp(firebaseConfig);
+}
+
+
+export const initializeFirestore = async (app: any) => {
+
+	const firestore = getFirestore(app);
+
+	if (config.FB_FIRESTORE_EMULATOR_HOST && config.FB_FIRESTORE_EMULATOR_PORT) {
+		if (FIRESTORE_EMULATOR_REQUIRED) {
+			connectFirestoreEmulator(firestore, config.FB_FIRESTORE_EMULATOR_HOST, +config.FB_FIRESTORE_EMULATOR_PORT);
+			FIRESTORE_EMULATOR_REQUIRED = false;
+		}
+		console.log('Using firestore emulator');
+	} else {
+		console.log('Using production firestore');
+	}
+
+	return firestore;
+}
+
+export const prepareData = async (data: any, calculationYear: string, currentDate: string, firestore: any) => {
```

Comment:
I would use the proper types instead of `any`, then we can profit more from type checking.

Reply:
Here I am not sure how to deal with JSON, if you could provide some help, it would be very much appreciated. I adapted in all the functions the types where I was sure :-)

Reply:
True, external data is always tricky.
I think you could create an interface for it. Then one can see in one glance what we expect from the request.

```
interface Data {
 myexpectecedVariable: string
}
```